python_modules/dagster/dagster/_core/definitions/parent_asset_graph_differ.py

```diff
@@ -0,0 +1,144 @@
+from dagster._core.errors import DagsterInvariantViolationError
+from dagster._core.host_representation import ExternalRepository
+from dagster._core.workspace.context import BaseWorkspaceRequestContext
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.events import (
+        AssetKey,
+    )
+
+
+class ChangeReason(Enum):
+    NEW = "NEW"
+    CODE_VERSION = "CODE_VERSION"
+    INPUTS = "INPUTS"
+
+
+def _get_external_repo_from_context(
+    context: BaseWorkspaceRequestContext, code_location_name: str, repository_name: str
+) -> Optional[ExternalRepository]:
+    """Returns the ExternalRepository specified by the code location name and repository name
+    for the provided workspace context. If the repository doesn't exist, return None.
+    """
+    if context.has_code_location(code_location_name):
+        cl = context.get_code_location(code_location_name)
+        if cl.has_repository(repository_name):
+            return cl.get_repository(repository_name)
+
+
+class ParentAssetGraphDiffer:
+    """Given two asset graphs, parent_asset_graph and branch_asset_graph, we can compute how the
+    assets in branch_asset_graph have changed with respect to parent_asset_graph. The ChangeReason
+    enum contains the list of potential changes an asset can undergo. If the parent_asset_graph is None,
+    this indicates that the branch_asset_graph does not yet exist in the parent deployment. In this case
+    we will consider every asset New.
+    """
+
+    _branch_asset_graph: Optional["ExternalAssetGraph"]
+    _branch_asset_graph_load_fn: Optional[Callable[[], "ExternalAssetGraph"]]
+    _parent_asset_graph: Optional["ExternalAssetGraph"]
+    _parent_asset_graph_load_fn: Optional[Callable[[], "ExternalAssetGraph"]]
+
+    def __init__(
+        self,
+        branch_asset_graph: Union["ExternalAssetGraph", Callable[[], "ExternalAssetGraph"]],
+        parent_asset_graph: Optional[
+            Union["ExternalAssetGraph", Callable[[], "ExternalAssetGraph"]]
+        ] = None,
+    ):
+        if parent_asset_graph is None:
+            # if parent_asset_graph is None, then the asset graph in the branch deployment does not exist
+            # in the parent deployment
+            self._parent_asset_graph = None
+            self._parent_asset_graph_load_fn = None
+        elif isinstance(parent_asset_graph, ExternalAssetGraph):
+            self._parent_asset_graph = parent_asset_graph
+            self._parent_asset_graph_load_fn = None
+        else:
+            self._parent_asset_graph = None
+            self._parent_asset_graph_load_fn = parent_asset_graph
+
+        if isinstance(branch_asset_graph, ExternalAssetGraph):
+            self._branch_asset_graph = branch_asset_graph
+            self._branch_asset_graph_load_fn = None
+        else:
+            self._branch_asset_graph = None
+            self._branch_asset_graph_load_fn = branch_asset_graph
+
+    @classmethod
+    def from_external_repositories(
```

Comment:
I'm not using the `__init__` to make this class anywhere anymore, but having this method for making the class is useful because I can return None if `parent_workspace` is None and we're not in a branch deployment. l

Still thinking on a way to potentially consolidate this...